### PR TITLE
test: add tests for composer.ts

### PR DESCRIPTION
**Pull request by KnorpelSenf**
Tests matter

### CODE REVIEW

File: test/composer.test.ts [hunk 1]
```diff
@@ -1,20 +1,57 @@
...
 
+describe("BotError", () => {
+    const ctx = new Context(
+        // deno-lint-ignore no-explicit-any
+        { message: { text: "test" } } as any,
+        // deno-lint-ignore no-explicit-any
+        0 as any,
+        // deno-lint-ignore no-explicit-any
+        0 as any,
+    );
```
[the remark this answers is not in the document]
Comment (KnorpelSenf) [reply test/composer.test.ts:24]: Fair enough, I'll change it

File: test/deps.test.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 export {
     readAll,
     readerFromIterable,
-} from "https://deno.land/std@0.178.0/streams/mod.ts";
+} from "https://deno.land/std@0.181.0/streams/mod.ts";
```
Comment (dcdunkan) [inline test/deps.test.ts:4]:
```suggestion
} from "https://deno.land/std@0.184.0/streams/mod.ts";
```

in other places too.

**Review by dcdunkan — CHANGES_REQUESTED** [review]

File: src/platform.deno.ts [hunk 3]
```diff
@@ -17,7 +17,7 @@ if (isDeno) {
 
 // === Export system-specific operations
 // Turn an AsyncIterable<Uint8Array> into a stream
-export { readableStreamFromIterable as itrToStream } from "https://deno.land/std@0.178.0/streams/mod.ts";
+export { readableStreamFromIterable as itrToStream } from "https://deno.land/std@0.181.0/streams/mod.ts";
```
Comment (dcdunkan) [inline src/platform.deno.ts:20]: Here as well? to 0.184.0

**Review by dcdunkan — APPROVED** [review]

File: src/platform.deno.ts (hunk 3 above)
Comment (KnorpelSenf) [reply src/platform.deno.ts:20]: Yes.
```suggestion
export { readableStreamFromIterable as itrToStream } from "https://deno.land/std@0.184.0/streams/mod.ts";
```

**Comment by KnorpelSenf** [comment]
Need to investigate why CI fails.

**Review by roziscoding — APPROVED** [review]

File: .vscode/settings.json [hunk 4]
```diff
@@ -1,8 +1,11 @@
...
-    "editor.codeActionsOnSave": { "source.fixAll": true },
+    "editor.codeActionsOnSave": {
+        "source.fixAll": true
+    },
     "deno.enable": true,
     "deno.lint": true,
     "deno.config": "./deno.jsonc",
-    "editor.defaultFormatter": "denoland.vscode-deno"
+    "editor.defaultFormatter": "denoland.vscode-deno",
+    "vue.features.codeActions.enable": false
```
Comment (KnorpelSenf) [inline .vscode/settings.json:10]: Please revert

Comment (roziscoding) [reply .vscode/settings.json:10]: 🤦🏻

File: src/core/client.ts [hunk 5]
```diff
@@ -269,7 +269,12 @@ class ApiClient<R extends RawApi> {
             ? createFormDataPayload(payload, (err) => streamErr.catch(err))
             : createJsonPayload(payload);
         const sig = controller.signal;
-        const options = { ...opts.baseFetchConfig, signal: sig, ...config };
+        const options = {
+            ...opts.baseFetchConfig,
+            // deno-lint-ignore no-explicit-any
```
Comment (EdJoPaTo) [inline src/core/client.ts:274]: Do not use `as any` and ignore the lint to work around a type issue. If it’s a type issue tell that directly to typescript via ts-ignore-error

File: test/composer.test.ts [hunk 6]
```diff
@@ -1,20 +1,51 @@
...
+        assertStringIncludes(e(42), "of type number");
+        assertStringIncludes(e(() => {}), "of type function");
+    });
+});
+
 describe("Composer", () => {
     let composer: Composer<Context>;
-    const ctx = { message: { text: "" } } as Context;
+    const ctx = new Context(
+        // deno-lint-ignore no-explicit-any
```
Comment (EdJoPaTo) [inline test/composer.test.ts:39]: Again, use ts-ignore-error over `as any` and a lint comment. (I will omit the other hints to this in this file)

File: test/composer.test.ts [hunk 7]
```diff
@@ -23,48 +54,435 @@ describe("Composer", () => {
 
     it("should call handlers", async () => {
         composer.use(middleware);
-        await composer.middleware()(ctx, next);
+        await exec();
```
Comment (EdJoPaTo) [inline test/composer.test.ts:57]: What’s the benefit of hiding this in the `exec` function?

**Review by EdJoPaTo — CHANGES_REQUESTED** [review]

File: src/core/client.ts (hunk 5 above)
Comment (KnorpelSenf) [reply src/core/client.ts:274]: This is not related to this PQ. I can fix it separately at a later point.

File: test/composer.test.ts (hunk 7 above)
Comment (KnorpelSenf) [reply test/composer.test.ts:57]: I expected `exec` to be rather complex so I did not want to repeat myself. But it's also kinda opaque, so I'm not happy either way. Suggestions? Inline `exec` everywhere?

File: src/core/client.ts (hunk 5 above)
Comment (EdJoPaTo) [reply src/core/client.ts]: This PQ adds the as any so I think it’s done in this one?

Comment (KnorpelSenf) [reply src/core/client.ts]: It no longer adds this. This is outdated by now.

File: package.json [hunk 8]
```diff
@@ -26,9 +26,9 @@
     "devDependencies": {
         "@types/debug": "^4.1.7",
         "@types/node": "^12.20.55",
-        "@types/node-fetch": "^2.6.2",
+        "@types/node-fetch": "2.6.2",
```
Comment (EdJoPaTo) [inline package.json:29]: Why is the requirement changed for the types but not for the dependency itself? Shouldn’t be both or neither changed?

Comment (KnorpelSenf) [reply package.json:29]: No. The types are broken so we pin them for us until they're fixed. That doesn't mean we cannot permit library users to use newer versions.

File: test/composer.test.ts (hunk 7 above)
Comment (KnorpelSenf) [reply test/composer.test.ts:57]: Is that upvote a yes to the inlining?

File: package.json [hunk 9]
```diff
@@ -26,9 +26,9 @@
...
         "all-contributors-cli": "^6.24.0",
-        "deno2node": "^1.7.2"
+        "deno2node": "^1.8.1"
```
Comment (dcdunkan) [inline package.json:31]: I guess the version needs to be updated in deno.jsonc backport task as well?

Comment (KnorpelSenf) [reply package.json:31]: Yes

**Pull request merged**